File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # # recursive 遞迴

#         # 結束條件
#         if not head or head.next is None:
#             return head
#         # head 跑到最後
#         new_head = self.reverseList(head.next)
#         # reverse
#         head.next.next = head
#         head.next = None

#         return new_head

        # iteration 迭代

        prev = None
        curr = head

        while curr is not None:
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode 
        
        return prev

File: 0206-reverse-linked-list/0206-reverse-linked-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
    // struct ListNode *prev_node = NULL;
    // struct ListNode *curr_node = head;
    // struct ListNode *next_node = NULL;

    // while (curr_node != NULL) {
    //     next_node = curr_node -> next;
    //     curr_node -> next = prev_node;
    //     prev_node = curr_node;
    //     curr_node = next_node;
    // }
    // return prev_node;

    struct ListNode* curr_node = head;
    struct ListNode* prev_node = NULL;
    struct ListNode* next_node = NULL;

    while (curr_node != NULL){
        next_node = curr_node -> next;
        curr_node -> next = prev_node;

        prev_node = curr_node;
        curr_node = next_node;
    }
    return prev_node;
}